src/Nethermind/Nethermind.Blockchain/Producers/DevBlockProducer.cs

```diff
@@ -46,25 +47,47 @@ public class DevBlockProducer : BaseBlockProducer
         public override void Start()
         {
             _txPool.NewPending += OnNewPendingTx;
+            BlockTree.NewHeadBlock += OnNewHeadBlock;
         }
-
+        
         public override async Task StopAsync()
         {
             _txPool.NewPending -= OnNewPendingTx;
+            BlockTree.NewHeadBlock -= OnNewHeadBlock;
             await Task.CompletedTask;
         }
 
         protected override UInt256 CalculateDifficulty(BlockHeader parent, UInt256 timestamp) => 1;
 
         private void OnNewPendingTx(object sender, TxEventArgs e)
         {
-            TryProduceNewBlock(CancellationToken.None).ContinueWith(t =>
+            OnNewPendingTxAsync(e);
+        }
+
+        private async void OnNewPendingTxAsync(TxEventArgs e)
+        {
+            _newBlockLock.Wait(TimeSpan.FromSeconds(1));
+            try
             {
-                if (t.IsFaulted)
+                if (!await TryProduceNewBlock(CancellationToken.None))
                 {
-                    if(Logger.IsError) Logger.Error($"Failed to produce block after receiving transaction {e.Transaction}", t.Exception);
+                    _newBlockLock.Release();
                 }
-            });
+            }
+            catch (Exception exception)
+            {
+                if (Logger.IsError) Logger.Error($"Failed to produce block after receiving transaction {e.Transaction}", exception);
+                _newBlockLock.Release();
```

Comment:
move Release to finally - at the moment if you succeed to produce the block it seems that you are not releasing the signal?

Reply:
is it released somewhere else?

Reply:
We are waiting with the release until the block is processed. If the processing fails we still will enter but we are giving 1 second timeout for the lock just for that.

This is NethDev only - good enough.

Reply:
Basically:
1. We acquire the lock, or wait 1 second
2. If we fail to produce block immediately - we release the lock
3. We wait for NewHead
4. If New head arrives we release the lock
5. If new head doesn't arrive (exception?) We have that 1sec threshold for that.

Its not perfect, but its good enough for testing purposes, which is the requirements for this producer.